File: android/app/src/main/java/org/findmyfam/services/PendingInviteStore.kt
package org.findmyfam.services

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.findmyfam.models.PendingInvite
import org.json.JSONArray
import org.json.JSONObject
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Persists pending group invites -- invites where the user has published
 * their key package but hasn't yet received a Welcome event.
 *
 * SharedPreferences-backed so pending state survives app restarts.
 */
@Singleton
class PendingInviteStore @Inject constructor(
    @ApplicationContext context: Context
) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("fmf_pending_invites", Context.MODE_PRIVATE)

    private val _pendingInvites = MutableStateFlow<List<PendingInvite>>(emptyList())
    val pendingInvites: StateFlow<List<PendingInvite>> = _pendingInvites.asStateFlow()

    init {
        load()
    }

    /**
     * Record a new pending invite.
     */
    fun add(invite: PendingInvite) {
        if (_pendingInvites.value.any { it.groupHint == invite.groupHint }) return
        _pendingInvites.value = _pendingInvites.value + invite
        save()
        Timber.i("PendingInviteStore: added invite for group ${invite.groupHint}")
    }

    /**
     * Remove a pending invite by group hint (e.g. when a Welcome is received).
     */
    fun remove(groupHint: String) {
        _pendingInvites.value = _pendingInvites.value.filter { it.groupHint != groupHint }
        save()
        Timber.i("PendingInviteStore: removed invite for group $groupHint")
    }

    /**
     * Remove all pending invites.
     */
    fun removeAll() {
        _pendingInvites.value = emptyList()
        save()
    }

    /**
     * Remove invites that match any of the given active group IDs.
     * Called after receiving Welcomes to clean up resolved invites.
     */
    fun removeResolved(activeGroupIds: Set<String>) {
        val before = _pendingInvites.value.size
        _pendingInvites.value = _pendingInvites.value.filter { it.groupHint !in activeGroupIds }
        val removed = before - _pendingInvites.value.size
        if (removed > 0) {
            save()
            Timber.i("PendingInviteStore: auto-removed $removed resolved invite(s)")
        }
    }

    // --- Persistence ---

    private fun save() {
        val arr = JSONArray()
        for (invite in _pendingInvites.value) {
            arr.put(JSONObject().apply {
                put("groupHint", invite.groupHint)
                put("inviterNpub", invite.inviterNpub)
                put("createdAt", invite.createdAt)
            })
        }
        prefs.edit().putString(STORAGE_KEY, arr.toString()).apply()
    }

    private fun load() {
        val json = prefs.getString(STORAGE_KEY, null) ?: return
        try {
            val arr = JSONArray(json)
            val list = mutableListOf<PendingInvite>()
            for (i in 0 until arr.length()) {
                val obj = arr.getJSONObject(i)
                list.add(
                    PendingInvite(
                        groupHint = obj.getString("groupHint"),
                        inviterNpub = obj.getString("inviterNpub"),
                        createdAt = obj.getLong("createdAt")
                    )
                )
            }
            _pendingInvites.value = list
        } catch (e: Exception) {
            Timber.w(e, "Failed to load pending invites")
        }
    }

    companion object {
        private const val STORAGE_KEY = "fmf.pendingInvites"
    }
}

File: Sources/Services/PendingInviteStore.swift
import Foundation
import WhistleCore

/// Persists pending group invites — invites where the user has published
/// their key package but hasn't yet received a Welcome event.
///
/// UserDefaults-backed so pending state survives app restarts.
@MainActor
final class PendingInviteStore: ObservableObject {

    private static let storageKey = "fmf.pendingInvites"

    @Published private(set) var pendingInvites: [PendingInvite] = []

    /// - Parameter skipLoad: When `true`, skip loading from UserDefaults (for testing).
    init(skipLoad: Bool = false) {
        if !skipLoad { load() }
    }

    // MARK: - Mutations

    /// Record a new pending invite.
    func add(_ invite: PendingInvite) {
        // Don't duplicate
        guard !pendingInvites.contains(where: { $0.groupHint == invite.groupHint }) else { return }
        pendingInvites.append(invite)
        save()
        FMFLogger.marmot.info("PendingInviteStore: added invite for group \(invite.groupHint)")
    }

    /// Remove a pending invite by group hint (e.g. when a Welcome is received).
    func remove(groupHint: String) {
        pendingInvites.removeAll { $0.groupHint == groupHint }
        save()
        FMFLogger.marmot.info("PendingInviteStore: removed invite for group \(groupHint)")
    }

    /// Remove all pending invites.
    func removeAll() {
        pendingInvites.removeAll()
        save()
    }

    /// Remove invites that match any of the given active group IDs.
    /// Called after receiving Welcomes to clean up resolved invites.
    func removeResolved(activeGroupIds: Set<String>) {
        let before = pendingInvites.count
        pendingInvites.removeAll { activeGroupIds.contains($0.groupHint) }
        let removed = before - pendingInvites.count
        if removed > 0 {
            save()
            FMFLogger.marmot.info("PendingInviteStore: auto-removed \(removed) resolved invite(s)")
        }
    }

    // MARK: - Persistence

    private func save() {
        guard let data = try? JSONEncoder().encode(pendingInvites) else { return }
        UserDefaults.standard.set(data, forKey: Self.storageKey)
    }

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: Self.storageKey),
              let decoded = try? JSONDecoder().decode([PendingInvite].self, from: data) else {
            return
        }
        pendingInvites = decoded
    }
}
